core/src/main/java/org/elasticsearch/bootstrap/Spawner.java

```diff
@@ -0,0 +1,125 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.bootstrap;
+
+import org.apache.lucene.util.Constants;
+import org.elasticsearch.env.Environment;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Spawns native plugin controller processes if present.  Will only work prior to seccomp being set up.
+ */
+final class Spawner implements Closeable {
+
+    private static final String PROGRAM_NAME = Constants.WINDOWS ? "controller.exe" : "controller";
+    private static final String PLATFORM_NAME = makePlatformName(Constants.OS_NAME, Constants.OS_ARCH);
+    private static final String TMP_ENVVAR = "TMPDIR";
+
+    /**
+     * References to the processes that have been spawned, so that we can destroy them.
+     */
+    private final List<Process> processes = new ArrayList<>();
+
+    @Override
+    public void close() throws IOException {
+        for (Process process : processes) {
+            process.destroy();
+        }
+        processes.clear();
+    }
+
+    /**
+     * For each plugin, attempt to spawn the controller daemon.  Silently ignore any plugins
+     * that don't include a controller for the correct platform.
+     */
+    void spawnNativePluginControllers(Environment environment) throws IOException {
+        if (Files.exists(environment.pluginsFile())) {
+            try (DirectoryStream<Path> stream = Files.newDirectoryStream(environment.pluginsFile())) {
+                for (Path plugin : stream) {
+                    Path spawnPath = makeSpawnPath(plugin);
+                    if (Files.isRegularFile(spawnPath)) {
+                        spawnNativePluginController(spawnPath, environment.tmpFile());
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Attempt to spawn the controller daemon for a given plugin.  The spawned process
+     * will remain connected to this JVM via its stdin, stdout and stderr, but the
+     * references to these streams are not available to code outside this package.
+     */
+    private void spawnNativePluginController(Path spawnPath, Path tmpPath) throws IOException {
+        ProcessBuilder pb = new ProcessBuilder(spawnPath.toString());
+
+        // The only environment variable passes on the path to the temporary directory
+        pb.environment().clear();
+        pb.environment().put(TMP_ENVVAR, tmpPath.toString());
```

Comment:
the reason why I suggested to make it configurable is that we could pass in our own values in tests that's all... not a big deal